baseapp/baseapp.go

```diff
@@ -714,7 +760,7 @@ func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*s
 	// NOTE: GasWanted is determined by the AnteHandler and GasUsed by the GasMeter.
 	for i, msg := range msgs {
 		// skip actual execution for (Re)CheckTx mode
-		if mode == runTxModeCheck || mode == runTxModeReCheck {
+		if mode == runTxModeCheck || mode == runTxModeReCheck || mode == runTxPrepareProposal || mode == runTxProcessProposal {
```

Comment:
Mhhh I think we need to think very carefully now about the tx execution flow.

My thinking is, it's now possible for application's to completely control `ProcessProposal`. This means that they can execute the _entire_ block, i.e. all the txs. Which means we need to re-think slightly how `runMsgs` and `runTx` behave.

However, I don't want to muddle this PR. So I propose the following:

1. Simply this to `mode != runTxModeDeliver`
2. Create an issue linking to this comment
3. I'll open a separate PR that addresses it.

Reply:
Yea `runTx` and `runMsg` taking "mode" as a parameter has been quite confusing to develop and debug.  I think we'd like to refactor it, but not wanting to increase the scope here we've continued in the same "mode"-driven paradigm.

But there are now 6 modes:
https://github.com/cosmos/cosmos-sdk/blob/8d4e97b9aae77a4979d68ae4ec68f5009e0c1b95/baseapp/baseapp.go#L28-L35

Since `mode != runTxModeDeliver` is not an inversion of the same logic, it is causing test failures.  The inverse of the statement as is would be

```
		if mode != runTxModeDeliver && mode != runTxModeSimulate {
```

Does that seem like an improvement to you?